test/exchangeIssuance/exchangeIssuance-staged.spec.ts

```diff
@@ -0,0 +1,906 @@
+  return await getIssueSetForExactETH(setToken, ethInput, uniswapRouter, weth);
+};
+
+const getIssueExactSetFromETH = async (setToken: SetToken, amountSet: BigNumber, uniswapRouter: UniswapV2Router02, weth: string) => {
+  const components = await setToken.getComponents();
+  let sumEth = BigNumber.from(0);
+  for (let i = 0; i < components.length; i++) {
+    const componentAmount = amountSet.mul(await setToken.getDefaultPositionRealUnit(components[i])).div(ether(1));
+    const ethAmount = (await uniswapRouter.getAmountsIn(componentAmount, [weth, components[i]]))[0];
+    sumEth = sumEth.add(ethAmount);
+  }
+  return sumEth;
+};
+
+const getIssueExactSetFromToken = async (setToken: SetToken, inputToken: StandardTokenMock, inputAmount: BigNumber,
+  amountSet: BigNumber, uniswapRouter: UniswapV2Router02, weth: string) => {
+
+  const ethCost = await getIssueExactSetFromETH(setToken, amountSet, uniswapRouter, weth);
+  const inputEthValue = (await uniswapRouter.getAmountsOut(inputAmount, [inputToken.address, weth]))[1];
+  const refundAmount = inputEthValue.sub(ethCost);
+
+  return refundAmount;
+};
+
+const getRedeemExactSetForETH = async (setToken: SetToken, amountSet: BigNumber, uniswapRouter: UniswapV2Router02, weth: string) => {
+  const components = await setToken.getComponents();
+  let sumEth = BigNumber.from(0);
+  for (let i = 0; i < components.length; i++) {
+    const componentAmount = amountSet.mul(await setToken.getDefaultPositionRealUnit(components[i])).div(ether(1));
+    const ethAmount = (await uniswapRouter.getAmountsOut(componentAmount, [components[i], weth]))[1];
+    sumEth = sumEth.add(ethAmount);
+  }
+  return sumEth;
+};
+
+const getRedeemExactSetForToken = async (setToken: SetToken, outputToken: StandardTokenMock,
+  amountSet: BigNumber, uniswapRouter: UniswapV2Router02, weth: string) => {
+
+  const ethOut = await getRedeemExactSetForETH(setToken, amountSet, uniswapRouter, weth);
+  const tokenOut = (await uniswapRouter.getAmountsOut(ethOut, [weth, outputToken.address]))[1];
+  return tokenOut;
+};
+
+describe("ExchangeIssuance", async () => {
+  let owner: Account;
+  let user: Account;
+  let setV2Setup: SetFixture;
+
+  let deployer: DeployHelper;
+  let setToken: SetToken;
+
+  let exchangeIssuance: ExchangeIssuance;
+
+  before(async () => {
+    [
+      owner,
+      user,
+    ] = await getAccounts();
+
+    deployer = new DeployHelper(owner.wallet);
+
+    setV2Setup = getSetFixture(owner.address);
+    await setV2Setup.initialize();
+
+    setToken = await setV2Setup.createSetToken(
+      [setV2Setup.dai.address, setV2Setup.wbtc.address],
+      [ether(0.5), BigNumber.from(10).pow(8)],
+      [setV2Setup.issuanceModule.address, setV2Setup.streamingFeeModule.address]
+    );
+    setV2Setup.issuanceModule.initialize(setToken.address, ADDRESS_ZERO);
+  });
+
+  addSnapshotBeforeRestoreAfterEach();
+
+  describe("#constructor", async () => {
```

Comment:
Something to also note here as you familiarize yourself with testing setup: you're deploying a SetToken for every test in the suite, but the constructor function won't be using it for example.